versioned/mongodb/src/main/java/com/dremio/nessie/versioned/store/mongodb/BsonConditionVisitor.java

```diff
@@ -0,0 +1,127 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.dremio.nessie.versioned.store.mongodb;
+
+import java.security.InvalidParameterException;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.bson.BsonDocument;
+import org.bson.conversions.Bson;
+import org.bson.internal.Base64;
+
+import com.dremio.nessie.versioned.impl.condition.ConditionExpression;
+import com.dremio.nessie.versioned.impl.condition.ConditionExpressionVisitor;
+import com.dremio.nessie.versioned.impl.condition.ExpressionFunction;
+import com.dremio.nessie.versioned.impl.condition.ExpressionPath;
+import com.dremio.nessie.versioned.impl.condition.Value;
+import com.dremio.nessie.versioned.impl.condition.ValueVisitor;
+import com.dremio.nessie.versioned.store.Entity;
+
+/**
+ * This class allows retrieval of ConditionExpression objects in BSON format.
+ */
+class BsonConditionVisitor implements ConditionExpressionVisitor<Bson> {
+  /**
+   * This provides a separation of queries on @{ExpressionFunction} from the object itself.
+   * This uses the Visitor design pattern to retrieve object attributes.
+   */
+  private static class BsonValueVisitor implements ValueVisitor<String> {
+    @Override
+    public String visit(Value value) {
+      return toMongoExpr(value.getValue());
+    }
+
+    @Override
+    public String visit(ExpressionFunction value) {
+      final ExpressionFunction.FunctionName name = value.getName();
+      final List<Value> arguments = value.getArguments();
+      if (arguments.size() != name.getArgCount()) {
+        throw new InvalidParameterException(
+          String.format("Number of arguments provided [%d] does not match the number expected [%d] for %s.",
+            arguments.size(), name.getArgCount(), name));
+      }
+
+      switch (name) {
+        case EQUALS:
+          // Special case SIZE, as the object representation is awkward for how it should be represented by MongoDB.
+          if (isSize(arguments.get(0))) {
+            return String.format("{%s: {\"$size\": %s}}",
+              arguments.get(0).getFunction().getArguments().get(0).accept(this), arguments.get(1).accept(this));
+          }
+
+          return String.format("{%s: %s}", arguments.get(0).accept(this), arguments.get(1).accept(this));
+        default:
+          throw new UnsupportedOperationException(String.format("%s is not a supported top-level MongoDB function.", name));
+      }
+    }
+
+    @Override
+    public String visit(ExpressionPath value) {
+      return value.getRoot().accept(BsonPathVisitor.INSTANCE, true);
+    }
+
+    private boolean isSize(Value value) {
```

Comment:
isSizeFunction? It's a longer name but it indicates we're talking about the function, and not some size.